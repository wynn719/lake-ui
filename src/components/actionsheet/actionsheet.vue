<template>
  <lake-popup :show="show" position="bottom" :lock-scroll="true" @click-mask="onClose">
    <div class="lake-actionsheet">
      <div class="lake-actionsheet-wrapper">
        <div class="lake-actionsheet-title" v-if="title">
          <slot name="title">{{ title }}</slot>
        </div>
        <section class="lake-actionsheet-body">
          <slot name="body">
            <lake-cell
              class="lake-actionsheet-item"
              v-for="(item, index) in items"
              :key="index"
              @click.native.prevent.stop="onItemSelect(item)"
            >
              <template slot="content">
                <div class="lake-actionsheet-item-name">{{ item.name }}</div>
                <div class="lake-actionsheet-subitem-name" v-if="item.subName">{{ item.subName }}</div>
              </template>
            </lake-cell>
          </slot>
        </section>
        <div class="lake-actionsheet-operation" v-if="showCancelBtn">
          <lake-button class="lake-actionsheet-operation-cancel" size="normal" @click="onClose">{{
            cancelBtnText
          }}</lake-button>
        </div>
      </div>
    </div>
  </lake-popup>
</template>

<script>
import lakePopup from '../popup';
import lakeButton from '../button';
import { cell as lakeCell, cellGroup as lakeCellGroup } from '../cell';

export default {
  name: 'lake-actionsheet',
  components: { lakePopup, lakeCell, lakeCellGroup, lakeButton },
  props: {
    show: {
      type: Boolean,
      default: false,
    },
    maskClose: {
      type: Boolean,
      default: true,
    },
    title: {
      type: String,
      default: '',
    },
    items: {
      type: Array,
      default() {
        return [];
      },
    },
    showCancelBtn: {
      type: Boolean,
      default: true,
    },
    cancelBtnText: {
      type: String,
      default: '取消',
    },
  },
  data() {
    return {};
  },
  methods: {
    onItemSelect(item) {
      this.$emit('select', item);
      this.$emit('close');
    },
    onClose() {
      this.$emit('close');
    },
  },
};
</script>

<style lang="less">
@import '../../style/themes/default.less';
@import '../../style/common/mixins.less';

.lake-actionsheet {
  &-wrapper {
    background-color: @color-bg-body;
    overflow: hidden;
  }
  &-title {
    position: relative;
    text-align: center;
    padding: 12px 15px;
    font-weight: bold;
    font-size: 13px;
    color: @color-text-primary;
    .single-line();
    .border-1px-bottom();
  }
  &-body {
    max-height: 300px;
    min-height: 100px;
    overflow-y: auto;
  }
  &-item {
    text-align: center;
  }
  &-item-name {
    text-align: center;
    font-size: 15px;
    .single-line();
  }
  &-subitem-name {
    font-size: 12px;
    margin-top: 4px;
    color: @color-text-secondary;
    .single-line();
  }
  &-operation {
    margin-top: 6px;
  }
  &-operation-cancel {
    border: none;
    border-radius: 0;
    color: @color-text-secondary;
  }
}
</style>
