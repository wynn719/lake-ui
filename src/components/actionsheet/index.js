import actionsheet from './actionsheet.vue';

export default actionsheet;
