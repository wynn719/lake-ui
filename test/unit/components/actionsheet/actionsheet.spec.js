import { mount } from '@vue/test-utils';
import actionsheet from '@/components/actionsheet';

describe('actionsheet', () => {
  const items = [
    {
      name: 'action 1',
    },
    {
      name: 'action 2',
    },
    {
      name: 'action 3',
    },
  ];

  // basic test
  test('should render component and match snapshot', () => {
    const wrapper = mount(actionsheet);

    expect(wrapper.contains('.lake-actionsheet')).toBe(false);
    expect(wrapper.element).toMatchSnapshot();
  });

  // props test
  test('should render component with show props and match snapshot', () => {
    const wrapper = mount(actionsheet, {
      propsData: {
        show: true,
      },
    });

    expect(wrapper.contains('.lake-actionsheet')).toBe(true);
    expect(wrapper.element).toMatchSnapshot();
  });

  test('should render component with title props and match snapshot', () => {
    const wrapper = mount(actionsheet, {
      propsData: {
        show: true,
        title: 'test title',
      },
    });

    expect(wrapper.find('.lake-actionsheet-title').text()).toBe('test title');
    expect(wrapper.element).toMatchSnapshot();
  });

  test('should render component with items props and match snapshot', () => {
    const wrapper = mount(actionsheet, {
      propsData: {
        show: true,
        items,
      },
    });

    expect(wrapper.findAll('.lake-actionsheet-item').filter(item => item.text()).length).toBe(3);
    expect(wrapper.element).toMatchSnapshot();
  });

  test('should render component with showCancelBtn props and match snapshot', () => {
    const wrapper = mount(actionsheet, {
      propsData: {
        show: true,
        showCancelBtn: false,
      },
    });

    expect(wrapper.contains('.lake-actionsheet-operation-cancel')).toBe(false);
  });

  test('should render component with showCancelBtn props', () => {
    const wrapper = mount(actionsheet, {
      propsData: {
        show: true,
        cancelBtnText: 'be careful cancel',
      },
    });

    expect(wrapper.find('.lake-actionsheet-operation-cancel').text()).toBe('be careful cancel');
  });

  // event test
  test('should emit a select event with item params', () => {
    const wrapper = mount(actionsheet, {
      propsData: {
        show: true,
        items,
      },
    });
    const click = jest.fn(item => item);
    const close = jest.fn();

    wrapper.vm.$on('select', click);
    wrapper.vm.$on('close', close);
    wrapper.find('.lake-actionsheet-item').trigger('click');

    expect(click.mock.calls.length).toBe(1);
    expect(close.mock.calls.length).toBe(1);
    expect(click.mock.results[0].value).toEqual(items[0]);
  });

  test('should emit a close event when click close btn', () => {
    const wrapper = mount(actionsheet, {
      propsData: {
        show: true,
      },
    });
    const click = jest.fn();

    wrapper.vm.$on('close', click);
    wrapper.find('.lake-actionsheet-operation-cancel').trigger('click');

    expect(click.mock.calls.length).toBe(1);
  });

  // slot test
});
