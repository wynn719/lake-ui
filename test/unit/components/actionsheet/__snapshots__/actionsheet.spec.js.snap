// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actionsheet should render component and match snapshot 1`] = `
<div>
  <!---->
   
  <!---->
</div>
`;

exports[`actionsheet should render component with items props and match snapshot 1`] = `
<div>
  <div
    appear=""
    class="lake-mask"
    name="lake-fade"
  />
   
  <div
    appear=""
    class="lake-popup lake-popup-bottom"
    name="lake-slide"
  >
    <div
      class="lake-actionsheet"
    >
      <div
        class="lake-actionsheet-wrapper"
      >
        <!---->
         
        <section
          class="lake-actionsheet-body"
        >
          <div
            class="lake-actionsheet-item lake-cell"
          >
            <div
              class="lake-cell-ft"
            >
              <div
                class="lake-actionsheet-item-name"
              >
                action 1
              </div>
               
              <!---->
            </div>
          </div>
          <div
            class="lake-actionsheet-item lake-cell"
          >
            <div
              class="lake-cell-ft"
            >
              <div
                class="lake-actionsheet-item-name"
              >
                action 2
              </div>
               
              <!---->
            </div>
          </div>
          <div
            class="lake-actionsheet-item lake-cell"
          >
            <div
              class="lake-cell-ft"
            >
              <div
                class="lake-actionsheet-item-name"
              >
                action 3
              </div>
               
              <!---->
            </div>
          </div>
        </section>
         
        <div
          class="lake-actionsheet-operation"
        >
          <button
            class="lake-btn lake-actionsheet-operation-cancel lake-btn-default lake-btn-normal"
            type="button"
          >
            <!---->
              取消
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`actionsheet should render component with show props and match snapshot 1`] = `
<div>
  <div
    appear=""
    class="lake-mask"
    name="lake-fade"
  />
   
  <div
    appear=""
    class="lake-popup lake-popup-bottom"
    name="lake-slide"
  >
    <div
      class="lake-actionsheet"
    >
      <div
        class="lake-actionsheet-wrapper"
      >
        <!---->
         
        <section
          class="lake-actionsheet-body"
        />
         
        <div
          class="lake-actionsheet-operation"
        >
          <button
            class="lake-btn lake-actionsheet-operation-cancel lake-btn-default lake-btn-normal"
            type="button"
          >
            <!---->
              取消
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`actionsheet should render component with title props and match snapshot 1`] = `
<div>
  <div
    appear=""
    class="lake-mask"
    name="lake-fade"
  />
   
  <div
    appear=""
    class="lake-popup lake-popup-bottom"
    name="lake-slide"
  >
    <div
      class="lake-actionsheet"
    >
      <div
        class="lake-actionsheet-wrapper"
      >
        <div
          class="lake-actionsheet-title"
        >
          test title
        </div>
         
        <section
          class="lake-actionsheet-body"
        />
         
        <div
          class="lake-actionsheet-operation"
        >
          <button
            class="lake-btn lake-actionsheet-operation-cancel lake-btn-default lake-btn-normal"
            type="button"
          >
            <!---->
              取消
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
